src/Grpc.AspNetCore.Server/Internal/PercentEncodingHelpers.cs

```diff
@@ -0,0 +1,218 @@
+
+            static void Encode(Span<char> span, string s)
+            {
+                Span<byte> unicodeBytesBuffer = stackalloc byte[MaxUnicodeCharsReallocate * MaxUtf8BytesPerUnicodeChar];
+
+                var writePosition = 0;
+                for (var i = 0; i < s.Length; i++)
+                {
+                    var current = s[i];
+                    if (current > AsciiMaxValue)
+                    {
+                        // Get additional unicode characters
+                        var unicodeCharCount = GetUnicodeCharCount(s, i);
+
+                        var numberOfBytes = Encoding.UTF8.GetBytes(s.AsSpan(i, unicodeCharCount), unicodeBytesBuffer);
+
+                        for (var count = 0; count < numberOfBytes; count++)
+                        {
+                            EscapeAsciiChar(ref span, ref writePosition, (char)unicodeBytesBuffer[count]);
+                        }
+                        i += unicodeCharCount - 1;
+                    }
+                    else if (IsUnreservedCharacter(current))
+                    {
+                        span[writePosition++] = current;
+                    }
+                    else
+                    {
+                        EscapeAsciiChar(ref span, ref writePosition, current);
+                    }
+                }
+            }
+        }
+
+        private static void EscapeAsciiChar(ref Span<char> span, ref int writePosition, char current)
+        {
+            span[writePosition++] = '%';
+            span[writePosition++] = HexChars[current >> 4];
+            span[writePosition++] = HexChars[current & 15];
+        }
+
+        private static int GetUnicodeCharCount(string value, int currentIndex)
+        {
+            var unicodeCharCount = 0;
+            var maxSize = Math.Min(value.Length - currentIndex, MaxUnicodeCharsReallocate - 1); // Leave a character for possible low surrogate
+            for (; unicodeCharCount < maxSize && value[currentIndex + unicodeCharCount] > AsciiMaxValue; unicodeCharCount++)
+            {
+            }
+
+            if (char.IsHighSurrogate(value[currentIndex + unicodeCharCount - 1]))
+            {
+                if (unicodeCharCount < value.Length - currentIndex && char.IsLowSurrogate(value[currentIndex + unicodeCharCount]))
+                {
+                    // Last character is a high surrogate so check ahead to see if it is followed by a low surrogate and include
+                    unicodeCharCount++;
+                }
+            }
+
+            return unicodeCharCount;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsUnreservedCharacter(char c)
+        {
+            return ((PercentEncodingUnreservedBitField[c / 8] >> (c % 8)) & 1) != 0;
+        }
+
+        // TODO(JamesNK): Remove WIP decode once we are certain Uri.UnescapeData works correctly
+#if Decode
+        public static string PercentDecode(string value)
```

Comment:
Looks like we only need the percent decoding for grpc client (which currently cannot access classes from Grpc.AspNetCore.Server anyway) and so far Uri.UnescapeDataString seems to work just fine, so let's remove the PercentDecode method entirely (no reason to check in dead code) and only add it once we are sure we actually need it (add a TODO to https://github.com/grpc/grpc-dotnet/blob/23099fefd3bc0353580cf5c7600716a9ea8f4306/src/Grpc.NetCore.HttpClient/Internal/GrpcCall.cs#L434 or file an issue to double check).
    
If we decide to add a custom PercentDecode method, there should be some test cases for "PercentDecode" too, including making sure that implementation conforms with this part of spec: *"When decoding invalid values, implementations MUST NOT error or throw away the message. At worst, the implementation can abort decoding the status message altogether such that the user would received the raw percent-encoded form. Alternatively, the implementation can decode valid portions while leaving broken %-encodings as-is or replacing them with a replacement character (e.g., '?' or the Unicode replacement character)."*